src/python_testing/TC_OCC_3_1.py

```diff
@@ -1,129 +1,160 @@
+from mobly import asserts
+
+
+class TC_OCC_3_1(MatterBaseTest):
+    async def read_occ_attribute_expect_success(self, attribute):
+        cluster = Clusters.Objects.OccupancySensing
+        endpoint = self.matter_test_config.endpoint
+        return await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=attribute)
+
+    async def write_hold_time(self, hold_time: Optional[Any]) -> Status:
+        dev_ctrl = self.default_controller
+        node_id = self.dut_node_id
+        endpoint = self.matter_test_config.endpoint
+
+        cluster = Clusters.OccupancySensing
+        write_result = await dev_ctrl.WriteAttribute(node_id, [(endpoint, cluster.Attributes.HoldTime(hold_time))])
+        return write_result[0].Status
+
+    def desc_TC_OCC_3_1(self) -> str:
+        return "[TC-OCC-3.1] Primary functionality with server as DUT"
+
+    def steps_TC_OCC_3_1(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commission DUT to TH and obtain DUT attribute list.", is_commissioning=True),
+            TestStep(2, "Change DUT HoldTime attribute value to 10 seconds."),
+            TestStep(3, "Do not trigger DUT occupancy sensing for the period of HoldTime. TH reads Occupancy attribute from DUT."),
+            TestStep(4, "Trigger DUT occupancy sensing to change the occupancy state and start a timer."),
+            TestStep(5, "After 10 seconds, TH reads Occupancy attribute from DUT.")
+        ]
+        return steps
+
+    def pics_TC_OCC_3_1(self) -> list[str]:
+        pics = [
+            "OCC.S",
+        ]
+        return pics
+
+    # Sends and out-of-band command to the all-clusters-app
+    def write_to_app_pipe(self, command):
+        with open(self.app_pipe, "w") as app_pipe:
+            app_pipe.write(command + "\n")
+        # Delay for pipe command to be processed (otherwise tests are flaky)
+        time.sleep(0.001)
+
+    @async_test_body
+    async def test_TC_OCC_3_1(self):
+        hold_time = 10  # 10 seconds for occupancy state hold time
+
+        # CI app pipe id creation
+        self.app_pipe = "/tmp/chip_all_clusters_fifo_"
+        self.is_ci = self.check_pics("PICS_SDK_CI_ONLY")
+        if self.is_ci:
+            app_pid = self.matter_test_config.app_pid
+            if app_pid == 0:
+                asserts.fail("The --app-pid flag must be set when PICS_SDK_CI_ONLY is set.c")
+            self.app_pipe = self.app_pipe + str(app_pid)
+
+        self.step(1)  # Commissioning already done
+
+        self.step(2)
+
+        cluster = Clusters.OccupancySensing
+        attributes = cluster.Attributes
+        attribute_list = await self.read_occ_attribute_expect_success(attribute=attributes.AttributeList)
+
+        has_hold_time = attributes.HoldTime.attribute_id in attribute_list
+
+        if has_hold_time:
+            # write 10 as a HoldTime attribute
+            await self.write_single_attribute(cluster.Attributes.HoldTime(hold_time))

```

Comment:
Don't use 10, it may be out of bounds. Use the HoldTimeLimits.HoldTimeMin. Let's make the examples use a HoldTimeMin of 2.

I think the test is written racy.

Reply:
Used 10 for enough delay.